### PR TITLE
Rework core

### PR DESCRIPTION
**Pull request by aboeglin**

### CODE REVIEW

File: src/core.js [hunk 1]
```diff
@@ -1,99 +1,79 @@
...
+  } else {
+    instance = Node(newNode.type);
+  }
 
-  const getResolvers = () => ({
-    mountedResolvers,
-    unmountedResolvers,
+  Object.defineProperty(newNode, "_instance", {
+    value: instance,
+    configurable: true,
```
Comment (brekk) [inline src/core.js:21]: And `writable: false` here? So some asshat 👋 can't do myInstance._instance = 'butts'? 😁

File: src/core.js [hunk 2]
```diff
@@ -1,98 +1,137 @@
...
-        const child = subtree ? subtree.children[i] : undefined;
-        nr.context = { ...nr.context, ...nodeResolver.context };
-        return traverse(config, child, nr);
-      }),
-      __internal: { resolver: resolverRef, owner: nodeResolver.owner },
-    };
-  } else if (node === undefined || node === null) {
-    return { __internal: { resolver: resolverRef, owner: nodeResolver.owner } };
+  // We need that copy for the unmount, otherwise the tree is already mutated and we can't diff it anymore.
+  const oldChildren = vnode.children ? [...vnode.children] : null;
```
Comment (brekk) [inline src/core.js:58]: Progeria

File: src/core.js [hunk 3]
```diff
@@ -1,98 +1,137 @@
...
+        (newChild && newChild.type !== oldChild.type)
+      ) {
+        oldChild._instance.triggerUnmounted();
+      }
+    });
+  }
+
+  if (Array.isArray(vnode.children) && vnode.children.length > 0) {
+    vnode.children = vnode.children.map((n, i) => {
+      const oldChild = oldChildren[i];
```
Comment (brekk) [inline src/core.js:109]: Shouldn't this whole if block be inside the `if(oldChildren)` block?

File: src/core.js (hunk 2 above)
Comment (aboeglin) [reply src/core.js:51]: What do you mean by that ?

File: src/core.js (hunk 1 above)
Comment (aboeglin) [reply src/core.js]: Seems like it is the default. But I'm happy to make it explicit for sure !

File: src/core.js (hunk 3 above)
Comment (aboeglin) [reply src/core.js]: Makes a lot of sense. I just did it and tests still pass.

Comment (aboeglin) [reply src/core.js]: I figured it'd be better if children would always be an array anyways. So I got rid of that if altogether in the end !

File: src/core.js (hunk 1 above)
Comment (aboeglin) [reply src/core.js]: check

File: src/core.js (hunk 2 above)
Comment (brekk) [reply src/core.js]: Progeria is a disease which causes degenerative aging -- "old children"

**Pull request merged**